Utils/azuretools-core/src/com/microsoft/azuretools/ui/embeddedbrowser/JxBrowserUtil.java

```diff
@@ -22,42 +22,328 @@
+        boolean result = true;
+        try {
+            targetFileStream = new FileInputStream(targetFile);
+            digest = DigestUtils.md5Hex(targetFileStream).toUpperCase();
+            digestFileContent = FileUtils.readFileToString(digestFile, "UTF-8");
+            targetFileStream.close();
+        } catch (IOException e) {
+            log.warning(String.format("Fail to read file %s or digest %s", filePathName, digestFilePathName));
+            result = false;
+        } finally {
+            IOUtils.closeQuietly(targetFileStream);
+        }
+
+        if (digestFileContent.equals(digest) == false) {
+            log.warning(String.format("Check sum error for file %s and digest file %s", filePathName, digestFilePathName));
+            return false;
+        }
+
+        return result;
+    }
+
+    /**
+     * Download file from Azure and save to target path specified. Return true if download succeeds
+     * @param filePathName
+     * @return
+     */
+    private static Boolean downloadFromAzure(@NotNull Path filePathName) {
+        boolean result = true;
+        try {
+            URI cloudBlobUri = new URI(AZURE_BLOB_URI);
+            CloudBlobClient serviceClient = new CloudBlobClient(cloudBlobUri);
+            CloudBlobContainer container = serviceClient.getContainerReference("libcontainer1");
+            CloudBlockBlob blob = container.getBlockBlobReference(filePathName.getFileName().toString());
+            File downloadingFile = filePathName.toFile();
+            blob.downloadToFile(downloadingFile.getAbsolutePath());
         } catch (Exception e) {
-            throw new Exception("Fail to load JxBrowser or load URL: " + e.getMessage());
+            result = false;
+            log.warning("Fail to download file from Azure: " + e.getMessage());
+        }
+
+        return result;
+    }
+
+    /**
+     * Download file and check digest with retry. Return false if all retry fails
+     * @param threadPool
+     * @param filePathNames
+     * @param currentRetry
+     * @param waitTime
+     * @return
+     */
+    private static boolean downloadAndCheckDigest(@NotNull ExecutorService threadPool, @NotNull List<Path> filePathNames, int currentRetry, int waitTime) {
+        List<Path> digestPathNames = new ArrayList<>();
+        for (int i = 0; i < filePathNames.size(); i++) {
+            digestPathNames.add(Paths.get(FilenameUtils.removeExtension(filePathNames.get(i).toString()) + ".md5"));
         }
 
-        return (JComponent)browserView;
+        boolean result = true;
+        try {
+            int i = filePathNames.size() - 1;
+            for (; i >= 0; i--) {
+                if (checkFileDigest(filePathNames.get(i), digestPathNames.get(i)))  {
+                    filePathNames.remove(i);
+                    digestPathNames.remove(i);
+                }
+            }
+
+            if (filePathNames.isEmpty()) return true;
+
+            if (currentRetry >= MAX_RETRY_TIMES) {
```

Comment:
It might be better to decrease `retryLeft` or `token`

Reply:
Fixed.

---
In reply to: [157416539](https://github.com/Microsoft/azure-tools-for-java/pull/1296#discussion_r157416539) [](ancestors = 157416539)